tests/linearizability/model.go

```diff
@@ -93,121 +96,77 @@ var etcdModel = porcupine.Model{
 }
 
 func step(state EtcdState, request EtcdRequest, response EtcdResponse) (bool, EtcdState) {
-	if request.Key == "" {
-		panic("invalid request")
-	}
-	if state.Key == "" {
-		return true, initState(request, response)
+	if len(state.PossibleValues) == 0 {
+		state.Key = request.Key
+		if ok, v := initValueRevision(request, response); ok {
+			state.PossibleValues = append(state.PossibleValues, v)
+		}
+		return true, state
 	}
 	if state.Key != request.Key {
-		panic("Multiple keys not supported")
+		panic("multiple keys not supported")
 	}
-	switch request.Op {
-	case Get:
-		return stepGet(state, request, response)
-	case Put:
-		return stepPut(state, request, response)
-	case Delete:
-		return stepDelete(state, request, response)
-	default:
-		panic("Unknown operation")
+	if response.Err != nil {
+		for _, v := range state.PossibleValues {
+			newV, _ := stepValue(v, request)
+			state.PossibleValues = append(state.PossibleValues, newV)
+		}
+	} else {
+		var i = 0
+		for _, v := range state.PossibleValues {
+			newV, expectedResponse := stepValue(v, request)
+			if expectedResponse == response {
+				state.PossibleValues[i] = newV
+				i++
```

Comment:
Why not just like this? 

```suggestion
				state.PossibleValues = []valueRevision{newV}
				return  true, state
```

Reply:
Not needed, by comparing response vs expected response we will not allow revision to decrease.

Reply:
This might not be correct. It will return `true` when any expected response matches the actual response. How can you guarantee that all the historical entries' revision <= response.Revision?

Reply:
Let's work with an example, 

Assuming initial value: revision is `10: 1`, and there are 3 successive failed PUT operation, so the highest revision cached will be 4, thanks to code at line 101 ~ 102. But actually the real revision on server side is still 1.  If the 4th PUT operation succeeded, then the returned revision will be 2, obviously it breaks the rule that revision should never decrease.

Reply:
What you described is a correct history. Failed requests are those that client didn't response from. Maybe they succeeded maybe not, client didn't see the results. So in case you described, failed requests were just lost.

Note, goal of the model is to confirm that history up to some point is correct. It porcupine library which tries to find a order that model accepts as correct.    

> How can you guarantee that all the historical entries' revision [+1] <= response.Revision?

This is because revision in state never decreases. Please read `stepValue` function it will never result in lowering the revision. So when comparing response, response that lowered revision will never match. If no responses match, we will remove all possible values and return invalid state.